src/main/java/net/irisshaders/iris/shaderpack/parsing/BooleanParser.java

```diff
@@ -0,0 +1,150 @@
+			boolean compute(boolean value, Stack<Boolean> valueStack) {
+				return value & valueStack.pop();
+			}
+		}, OR {
+			@Override
+			boolean compute(boolean value, Stack<Boolean> valueStack) {
+				return value | valueStack.pop();
+			}
+		}, NOT {
+			@Override
+			boolean compute(boolean value, Stack<Boolean> valueStack) {
+				return !value;
+			}
+		}, OPEN;
+
+		boolean compute(boolean value, Stack<Boolean> valueStack) {
+			return value;
+		}
+	}
+
+	/**
+	 * parses the given expression
+	 * @param expression expression to parse
+	 * @param valueLookup lookup of shadow options
+	 * @return result of the expression, or true if there was an error
+	 */
+	public static boolean parse(String expression, OptionValues valueLookup) {
+		try {
+			String option = "";
+			Stack<OPERATION> operationStack = new Stack<>();
+			Stack<Boolean> valueStack = new Stack<>();
+			for (int i = 0; i < expression.length(); i++) {
+				char c = expression.charAt(i);
+				switch (c) {
+					case '!' -> operationStack.push(OPERATION.NOT);
+					case '&' -> {
+						// add value first, because this checks for preceding NOTs
+						if (!option.isEmpty()) {
+							valueStack.push(processValue(option, valueLookup, operationStack));
+							option = "";
+						}
+						// AND operators have priority, so add a bracket if it's the first AND
+						if (operationStack.isEmpty() || !operationStack.peek().equals(OPERATION.AND)) {
+							operationStack.push(OPERATION.OPEN);
+						}
+						i++;
+						operationStack.push(OPERATION.AND);
+					}
+					case '|' -> {
+						// add value first, because this checks for preceding NOTs
+						if (!option.isEmpty()) {
+							valueStack.push(processValue(option, valueLookup, operationStack));
+							option = "";
+						}
+						// if there was an AND before, that needs to be evaluated because it takes priority
+						if (!operationStack.isEmpty() && operationStack.peek().equals(OPERATION.AND)) {
+							evaluate(operationStack, valueStack, true);
+						}
+						i++;
+						operationStack.push(OPERATION.OR);
+					}
+					case '(' -> operationStack.push(OPERATION.OPEN);
+					case ')' -> {
+						// add value first, because this checks for preceding NOTs
+						if (!option.isEmpty()) {
+							valueStack.push(processValue(option, valueLookup, operationStack));
+							option = "";
+						}
+						// if there was an AND before, that needs to be evaluated because it added its own bracket
+						if (!operationStack.isEmpty() && operationStack.peek().equals(OPERATION.AND)) {
+							evaluate(operationStack, valueStack, true);
+						}
+						evaluate(operationStack, valueStack, true);
+					}
+					case ' ' -> {}
+					default -> option += c;
+				}
+			}
+			if (!option.isEmpty()) {
+				valueStack.push(processValue(option, valueLookup, operationStack));
+			}
+			evaluate(operationStack, valueStack, false);
+			boolean result = valueStack.pop();
+			if (!valueStack.isEmpty() || !operationStack.isEmpty()) {
+				Iris.logger.warn(
```

Comment:
a LOGGER for this class could make these log messages more specific (it tags the messages with a class then). This is up to you